Add protocol-level @available annotation inheritance to generated mocks

Keitaro0226 opened the pull request:
## Overview

Fixes #314

This draft PR addresses an issue where `@available` annotations on protocols are not inherited by the generated mock classes. This can cause compilation errors when the protocol uses types or members that are only available in specific OS versions.

## Problem

Currently, when generating mocks for protocols with `@available` annotations, Mockolo does not apply these availability constraints to the generated mock. As a result, using the mock can trigger compile-time errors when availability-constrained types or members are involved.

### Example

```swift
/// @mockable
@available(iOS 18.0, *)
protocol Foo: Sendable {
    func bar()
}
```

Incorrect mock (missing `@available`):

```swift
public final class FooMock: Foo, @unchecked Sendable { 
    // ❌ Missing @available
}
```

Correct mock:

```swift
@available(iOS 18.0, *)
public final class FooMock: Foo, @unchecked Sendable { 
}
```

## Solution

- Extract protocol-level `@available` attributes and apply them to generated mocks.

## Benefits

- Prevents availability-related compilation errors.
- Aligns mock generation with Swift’s availability system.

---

Open to feedback!

Keitaro0226 committed:
Add protocol-level @available annotation inheritance to generated mocks

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/mockolo?pullRequest=315) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/mockolo?pullRequest=315) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/mockolo?pullRequest=315) it.</sub>

sidepelican commented:
Any updates on this? When this draft PR becomes ready?

Keitaro0226 commented:
Sorry for the long silence. I got sidetracked and lost track of this.

I'll add tests and mark this PR as ready soon. Thanks for your patience.

Keitaro0226 committed:
Merge remote-tracking branch 'upstream/master' into fix-available-annotation-handling

Keitaro0226 committed:
Fix unintended formatting changes introduced during conflict resolution

Keitaro0226 committed:
Add tests for protocol-level @available inheritance on Sendable mocks

sidepelican commented on Tests/TestSendable/FixtureSendable.swift:
```diff
@@ -208,4 +208,40 @@
 }
+
+@Fixture enum availableSendableProtocol {
+    /// @mockable
+    @available(iOS 18.0, *)
+    public protocol Foo: Sendable {
+        func bar() -> String
+    }
```
I want to enforce a compilation error when `@available` isn't applied to a mock.

```suggestion
    @available(macOS 99.0, *)
    struct Bar {}
    
    /// @mockable
    @available(macOS 99.0, *)
    protocol Foo: Sendable {
        func bar() -> Bar
    }
```

Keitaro0226 replied:
Fixed in 2fa6c3d.
I updated the fixture to use an availability-gated return type (`Bar`) so the generated mock must inherit `@available` to compile.

sidepelican commented on Sources/MockoloFramework/Models/ParsedEntity.swift:
```diff
@@ -65,12 +65,17 @@ struct ResolvedEntity {
 
     func model() -> Model {
         let metadata = entity.metadata
+        // Combine protocol-level attributes with member-level attributes
```
```suggestion
```

The meaning feels a bit off and it’s redundant, so let's remove it.

Keitaro0226 replied:
Fixed in 3e51751.
Removed the redundant comment.

sidepelican commented on Tests/TestActor/FixtureActor.swift:
```diff
@@ -100,6 +100,9 @@
         }
 
+        @MainActor
+        /// @mockable
```
It's strange for `@mockable` to come along too.
How about just inheriting `available`?
You can change `EntityNode.attributesDescription` requirement to `var attributes: [String] { get }`.

Keitaro0226 replied:
Fixed in 7f3a38b.
I changed declaration attribute handling to keep syntax attributes as separate entries, so `@mockable` trivia no longer gets carried along with `@available`.

sidepelican commented on Tests/TestSendable/FixtureSendable.swift:
```diff
@@ -208,4 +208,40 @@
+                return ""
+            }
+        }
+    }
```
I need a test case for protocol inheritance scenarios:


```swift
@available(macOS 100.0, *)
struct Bar {}

@available(macOS 90.0, *)
protocol Foo {
}

/// @mockable
@available(macOS 100.0, *)
protocol Foo2: Foo {
  var bar: Bar { get set }
}
```

Keitaro0226 replied:
Fixed in 32dffb1.

sidepelican reviewed (changes requested):
Thanks for the PR. Overall it looks good, but I noticed a few minor details that I'd like you to address before we merge.

Keitaro0226 committed:
Add availability-gated Sendable fixture coverage

Keitaro0226 committed:
Remove redundant ParsedEntity attribute comment

Keitaro0226 committed:
Fix availability attribute extraction for mock generation

Keitaro0226 committed:
Add availability test for inherited protocols

Keitaro0226 commented:
Thank you for the review.
I’ve addressed the feedback and pushed the updates. I’d appreciate it if you could take another look when you have a chance.
Thank you.

sidepelican reviewed (approved):
Sorry for the late review.
It looks good. Thank you for your contribution!
